From d68e8fb08650c40b7100866019b6e5d08636e88f Mon Sep 17 00:00:00 2001
From: Yannick Fertre <yannick.fertre@foss.st.com>
Date: Fri, 30 Jun 2023 10:18:05 +0200
Subject: [PATCH 0407/1141] drm/bridge/synopsys: dsi: support of output
 rotation

Signed-off-by: Yannick Fertre <yannick.fertre@foss.st.com>
Change-Id: I03a2f9d1fa757b4a9f4f9bc2a5a942f64134c4d4
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/315738
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 drivers/gpu/drm/bridge/synopsys/dw-mipi-dsi.c | 49 ++++++++++++++-----
 1 file changed, 37 insertions(+), 12 deletions(-)

diff --git a/drivers/gpu/drm/bridge/synopsys/dw-mipi-dsi.c b/drivers/gpu/drm/bridge/synopsys/dw-mipi-dsi.c
index 6a6d25c41158..f1f1e69a33b1 100644
--- a/drivers/gpu/drm/bridge/synopsys/dw-mipi-dsi.c
+++ b/drivers/gpu/drm/bridge/synopsys/dw-mipi-dsi.c
@@ -251,6 +251,7 @@ struct dw_mipi_dsi {
 	u32 lanes;
 	u32 format;
 	unsigned long mode_flags;
+	int rotation;
 
 #ifdef CONFIG_DEBUG_FS
 	struct dentry *debugfs;
@@ -360,6 +361,12 @@ static int dw_mipi_dsi_host_attach(struct mipi_dsi_host *host,
 			return ret;
 	}
 
+	/* check if a rotation is required on panel */
+	ret = of_property_read_u32(bridge->of_node, "rotation", &dsi->rotation);
+	if (ret < 0)
+		/* fail to get rotation, set 0 by default */
+		dsi->rotation = 0;
+
 	return 0;
 }
 
@@ -695,11 +702,16 @@ static void dw_mipi_dsi_video_packet_config(struct dw_mipi_dsi *dsi,
 	 * DSI_VNPCR.NPSIZE... especially because this driver supports
 	 * non-burst video modes, see dw_mipi_dsi_video_mode_config()...
 	 */
-
-	dsi_write(dsi, DSI_VID_PKT_SIZE,
-		       dw_mipi_is_dual_mode(dsi) ?
-				VID_PKT_SIZE(mode->hdisplay / 2) :
-				VID_PKT_SIZE(mode->hdisplay));
+	if (dsi->rotation == 90 || dsi->rotation == 270)
+		dsi_write(dsi, DSI_VID_PKT_SIZE,
+			       dw_mipi_is_dual_mode(dsi) ?
+					VID_PKT_SIZE(mode->vdisplay / 2) :
+					VID_PKT_SIZE(mode->vdisplay));
+	else
+		dsi_write(dsi, DSI_VID_PKT_SIZE,
+			       dw_mipi_is_dual_mode(dsi) ?
+					VID_PKT_SIZE(mode->hdisplay / 2) :
+					VID_PKT_SIZE(mode->hdisplay));
 }
 
 static void dw_mipi_dsi_command_mode_config(struct dw_mipi_dsi *dsi)
@@ -741,9 +753,15 @@ static void dw_mipi_dsi_line_timer_config(struct dw_mipi_dsi *dsi,
 {
 	u32 htotal, hsa, hbp, lbcc;
 
-	htotal = mode->htotal;
-	hsa = mode->hsync_end - mode->hsync_start;
-	hbp = mode->htotal - mode->hsync_end;
+	if (dsi->rotation == 90 || dsi->rotation == 270) {
+		htotal = mode->vtotal;
+		hsa = mode->vsync_end - mode->vsync_start;
+		hbp = mode->vtotal - mode->vsync_end;
+	} else {
+		htotal = mode->htotal;
+		hsa = mode->hsync_end - mode->hsync_start;
+		hbp = mode->htotal - mode->hsync_end;
+	}
 
 	/*
 	 * TODO dw drv improvements
@@ -764,10 +782,17 @@ static void dw_mipi_dsi_vertical_timing_config(struct dw_mipi_dsi *dsi,
 {
 	u32 vactive, vsa, vfp, vbp;
 
-	vactive = mode->vdisplay;
-	vsa = mode->vsync_end - mode->vsync_start;
-	vfp = mode->vsync_start - mode->vdisplay;
-	vbp = mode->vtotal - mode->vsync_end;
+	if (dsi->rotation == 90 || dsi->rotation == 270) {
+		vactive = mode->hdisplay;
+		vsa = mode->hsync_end - mode->hsync_start;
+		vfp = mode->hsync_start - mode->hdisplay;
+		vbp = mode->htotal - mode->hsync_end;
+	} else {
+		vactive = mode->vdisplay;
+		vsa = mode->vsync_end - mode->vsync_start;
+		vfp = mode->vsync_start - mode->vdisplay;
+		vbp = mode->vtotal - mode->vsync_end;
+	}
 
 	dsi_write(dsi, DSI_VID_VACTIVE_LINES, vactive);
 	dsi_write(dsi, DSI_VID_VSA_LINES, vsa);
-- 
2.39.2

